src/GraphQL.Tests/DI.xUnit/PrepareDependencyInjectionAttribute.cs

```diff
@@ -0,0 +1,71 @@
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Reflection;
+using GraphQL.DI;
+using GraphQL.Execution;
+using GraphQL.Types;
+using Xunit.Sdk;
+
+namespace GraphQL.Tests.DI;
+
+/// <summary>
+/// Allows test methods from <see cref="QueryTestBase{TSchema, TDocumentBuilder}"/> and descendants
+/// work with different DI providers. Mark your test method with [Theory] and [DependencyInjectionData]
+/// attributes instead of [Fact] attribute. Also add 'string name' parameter (unused, the name of DI provider)
+/// into your test method. By default all methods from <see cref="QueryTestBase{TSchema, TDocumentBuilder}"/>
+/// and descendants (not marked with [DependencyInjectionData]) work with the first DI provider from
+/// <see cref="DependencyInjectionAdapterData"/>.
+/// </summary>
+internal sealed class PrepareDependencyInjectionAttribute : BeforeAfterTestAttribute
```

Comment:
This class looks really weird. How does it line up the names to the instances? Also I see a static concurrent dictionary holding references to every DI service collection for every test and every DI provider.  Isn’t there another way to make this work? Can we just use async local to temporarily hold the service provider for the specific test in question and eliminate the dictionary and only instantiate the specific instance needed? And what if we want to test multiple serializers with multiple DI providers - how is that done?

Reply:
Maybe we should just use something a bit more conventional and simply set a property on the query test base like “DIProviderName = diName;” so it’s more flexible when writing “theory” tests.

Reply:
> This class looks really weird.

Agree.

> How does it line up the names to the instances?

Nohow. We just should use the correct order of adapters and their names in `DependencyInjectionAdapterData`.

>  Isn’t there another way to make this work? 

Maybe, but I don't know.

> Can we just use async local to temporarily hold the service provider for the specific test in question and eliminate the dictionary and only instantiate the specific instance needed?

It was my first attempt. Alas xUnit doesn't use the same async-local context.

> And what if we want to test multiple serializers with multiple DI providers - how is that done?

I thought about that. I don't know yet. Maybe some further tricks.

Reply:
> Maybe we should just use something a bit more conventional and simply set a property on the query test base like “DIProviderName = diName;” so it’s more flexible when writing “theory” tests.

Perhaps but I don't know. Again my goal was to integrate real DI containers with as little test changes as possible. We may continue further when writing other DI adapters implementations.

Reply:
I'd like to suggest that we have a goal of removing this class or rewriting it to work better.  But we can merge in the changes now if you like.

```suggestion
// TODO: remove or rewrite this class; should support theories in conjuction with multiple DI providers
internal sealed class PrepareDependencyInjectionAttribute : BeforeAfterTestAttribute
```

Reply:
> If there is a test that is a theory, and does not use DependencyInjectionAdapterData (the theory is for other data), is there any issue here? Will it still use just the first DI provider even if multiple are configured?

It uses the first provider, real provider, MSDI.